File: OpenMP_Samples/Answer/Reduction/C/reduction.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <omp.h>

#define  N           100
#define  MAX_ITER    100000
#define  MAX_THREADS 100


/* Please define the matrices in here */
static double  A[N];


void Reduction1(double A[N], int n); 
void Reduction2(double A[N], int n, int ib, double dtemp_t[MAX_THREADS], 
		int istart[MAX_THREADS], int iend[MAX_THREADS]); 
void Reduction3(double A[N], int n); 

int main(int argc, char* argv[]) {

     double  t0, t1, t2, t_w;
     double  dc_inv;
     double  dtemp, dtemp_t[MAX_THREADS];
     int     i, k;      
     int     ib, istart[MAX_THREADS], iend[MAX_THREADS];

     /* matrix generation --------------------------*/
     srand(1);
     dc_inv = 1.0/(double)RAND_MAX;
     for(i=0; i<N; i++) {
       A[i] = rand()*dc_inv;
     }
     /* end of matrix generation --------------------------*/

     /* Start of reduction1 routine ----------------------------*/
     t1 = omp_get_wtime();

     for(i=0; i<MAX_ITER; i++) {
       Reduction1(A, N);
     }

     t2 = omp_get_wtime();
     t_w =  t2 - t1; 
     /* End of reduction1 routine --------------------------- */

     printf("N  = %d \n",N);
     printf("MAX_ITER  = %d \n",MAX_ITER);
     printf("Reduction1 time  = %lf [sec.] \n",t_w);


     /* Start of reduction2 routine ----------------------------*/

     /* Set control information  */
     ib = N / omp_get_max_threads();
     for (k=0; k<omp_get_max_threads(); k++) {
       dtemp_t[k] = 0.0;
       istart[k] = ib*k;
       iend[k] = (k+1) * ib;
     }
     iend[omp_get_max_threads()-1] = N;

     t1 = omp_get_wtime();

     for(i=0; i<MAX_ITER; i++) {
       Reduction2(A, N, ib, dtemp_t, istart, iend);
     }

     t2 = omp_get_wtime();
     t_w =  t2 - t1; 
     /* End of reduction2 routine --------------------------- */

     printf("N  = %d \n",N);
     printf("MAX_ITER  = %d \n",MAX_ITER);
     printf("Reduction2 time  = %lf [sec.] \n",t_w);


     /* Start of reduction3 routine ----------------------------*/
     t1 = omp_get_wtime();

#pragma omp parallel for
     for(i=0; i<MAX_ITER; i++) {
       Reduction3(A, N);
     }

     t2 = omp_get_wtime();
     t_w =  t2 - t1; 
     /* End of reduction3 routine --------------------------- */

     printf("N  = %d \n",N);
     printf("MAX_ITER  = %d \n",MAX_ITER);
     printf("Reduction3 time  = %lf [sec.] \n",t_w);


     return 0;
}

void Reduction1(double A[N], int n) 
{
     int  i;
     double dtemp;

     dtemp = 0.0;
#pragma omp parallel for
     for(i=0; i<n; i++) {
       dtemp += A[i] * A[i];
     }
   
}

void Reduction2(double A[N], int n, int ib, double dtemp_t[MAX_THREADS], 
          int istart[MAX_THREADS], int iend[MAX_THREADS]) 
{
     int  i, k;
     double dtemp;

#pragma omp parallel for private(i)
     for (k=0; k<omp_get_max_threads(); k++ ) {
       for(i=istart[k]; i<iend[k]; i++) {
         dtemp_t[k] += A[i] * A[i];
       }
     }

     dtemp = 0.0;
     for (k=0; k<omp_get_max_threads(); k++ ) {
       dtemp += dtemp_t[k];
     }   
    
}

void Reduction3(double A[N], int n) 
{
     int  i;
     double dtemp;

     dtemp = 0.0;
     for(i=0; i<n; i++) {
       dtemp += A[i] * A[i];
     }
   
}

File: OpenMP_Samples/Answer/Reduction/F/reduction.f
       program main
       use omp_lib
       include 'reduction.inc'

       double precision,allocatable,dimension(:):: A

       double precision  t0, t1, t2, t_w
       double precision  dtemp
       double precision  dtemp_t(0:MAX_THREADS)  

       integer istart(0:MAX_THREADS),iend(0:MAX_THREADS)         
       integer  i, k, ib

       allocate (A(NN))

c      === matrix generation --------------------------
       call RANDOM_SEED
       do i=1, NN
         call RANDOM_NUMBER(dtemp)   
       enddo
c      === end of matrix generation ------------------------

c      === Start of routine1 ----------------------------
       t1 = omp_get_wtime()

       do i=1, MAX_ITER
         call Reduction1(A, NN)
       enddo

       t2 = omp_get_wtime()
       t_w =  t2 - t1 
c      === End of routine1 ---------------------------

       print *, "NN  = ", NN
       print *, "MAX_ITER  = ", MAX_ITER
       print *, "Reduction1 time[sec.] = ", t_w


c      === Start of routine2 ----------------------------
       ib = NN / omp_get_max_threads()
       do i=0, omp_get_max_threads()-1
         dtemp_t(i) = 0.0d0
         istart(i) = 1 + ib*i
         iend(i) = (i+1)*ib
       enddo
       iend(omp_get_max_threads()-1) = NN

       t1 = omp_get_wtime()

       do i=1, MAX_ITER
         call Reduction2(A, NN, ib, dtemp_t, istart, iend)
       enddo

       t2 = omp_get_wtime()
       t_w =  t2 - t1 
c      === End of routine2 ---------------------------

       print *, "NN  = ", NN
       print *, "MAX_ITER  = ", MAX_ITER
       print *, "Reduction2 time[sec.] = ", t_w

c      === Start of routine3 ----------------------------
       t1 = omp_get_wtime()

!$omp parallel do
       do i=1, MAX_ITER
         call Reduction3(A, NN)
       enddo
!$omp end parallel do

       t2 = omp_get_wtime()
       t_w =  t2 - t1 
c      === End of routine3 ---------------------------

       print *, "NN  = ", NN
       print *, "MAX_ITER  = ", MAX_ITER
       print *, "Reduction3 time[sec.] = ", t_w




c      -----------------------------------------

       deallocate(A)

       stop
       end



       subroutine Reduction1(A, n)
       use omp_lib
       include 'reduction.inc'

       double precision A(NN)
       integer n

       integer  i
       double precision dtemp

       dtemp = 0.0d0
!$omp parallel do reduction(+: dtemp)  
       do i=1, n
         dtemp = dtemp + A(i) * A(i)    
       enddo
!$omp end parallel do

       return
       end


       subroutine Reduction2(A, n, ib, dtemp_t, istart, iend)
       use omp_lib
       include 'reduction.inc'

       double precision A(NN)
       integer n, ib
       double precision dtemp_t(MAX_THREADS)
       integer istart(MAX_THREADS), iend(MAX_THREADS) 

       integer i, k
       double precision dtemp

!$omp parallel do private(i)  
       do k=0, omp_get_max_threads()-1
         do i=istart(k), iend(k)
           dtemp_t(k) = dtemp + A(i) * A(i)    
         enddo
       enddo 
!$omp end parallel do
      
       dtemp = 0.0d0
       do k=0, omp_get_max_threads()-1
          dtemp = dtemp + dtemp_t(k)
       enddo

       return
       end


       subroutine Reduction3(A, n)
       use omp_lib
       include 'reduction.inc'

       double precision A(NN)
       integer n

       integer  i
       double precision dtemp

       dtemp = 0.0d0
       do i=1, n
         dtemp = dtemp + A(i) * A(i)    
       enddo

       return
       end
